msal/src/Microsoft.Identity.Client/Internal/Requests/NonInteractive/UsernamePasswordRequest.cs

```diff
@@ -85,14 +84,22 @@ private async Task FetchAssertionFromWsTrustAsync()
                         userRealmResponse,
                         (cloudAudience, trustAddress, userName) =>
                         {
-                            return WsTrustRequestBuilder.BuildMessage(cloudAudience, trustAddress, (IWAInput)userName);
+                            return WsTrustRequestBuilder.BuildMessage(cloudAudience, trustAddress, (UsernamePasswordInput)userName);
                         }).ConfigureAwait(false);
 
                     // We assume that if the response token type is not SAML 1.1, it is SAML 2
                     userAssertion = new UserAssertion(
                         wsTrustResponse.Token,
                         (wsTrustResponse.TokenType == WsTrustResponse.Saml1Assertion) ? OAuth2GrantType.Saml11Bearer : OAuth2GrantType.Saml20Bearer);
                 }
+                else if (AuthenticationRequestParameters.Authority.AuthorityType == Core.Instance.AuthorityType.Aad)
+                {
+                    // handle grant flow
+                    if (!this.usernamePasswordInput.HasPassword())
+                    {
+                        throw new MsalException(MsalError.PasswordRequiredForManagedUserError);
```

Comment:
We should always throw MsalClientException or MsalServiceException. We expose MsalException as a base class as a convenience for users to catch and treat both at the same time, but we should never throw it.

Reply:
Seems like this should be a client exception as Bogdan suggests. Same is in IWarRequest

---
In reply to: [219454654](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1261#discussion_r219454654) [](ancestors = 219454654)